hook/wrapper.go

```diff
@@ -169,110 +170,94 @@ func NewScriptWrapper(opts ...scriptWrapperOpt) (*ScriptWrapper, error) {
-		return nil, err
+	var tmpl *template.Template
+	switch {
+	case isWindows && !isPOSIXHook && !isPwshHook:
+		tmpl = batchWrapperTmpl
+	case isWindows && isPwshHook:
+		tmpl = powershellWrapperTmpl
+	default:
+		tmpl = PosixShellWrapperTmpl
 	}
-	defer wrap.scriptFile.Close()
 
-	// We'll pump the ENV before the hook into this temp file
-	wrap.beforeEnvFile, err = shell.TempFileWithExtension(
-		"buildkite-agent-bootstrap-hook-env-before",
-	)
+	wrap.beforeEnvPath, err = tmpfile.KeepExtensionAndClose(hookWrapperDir, "hook-before-env")
 	if err != nil {
 		return nil, err
 	}
-	if err := wrap.beforeEnvFile.Close(); err != nil {
-		return nil, fmt.Errorf("failed to close before env file: %w", err)
-	}
 
-	// We'll then pump the ENV _after_ the hook into this temp file
-	wrap.afterEnvFile, err = shell.TempFileWithExtension(
-		"buildkite-agent-bootstrap-hook-env-after",
-	)
+	wrap.afterEnvPath, err = tmpfile.KeepExtensionAndClose(hookWrapperDir, "hook-after-env")
 	if err != nil {
 		return nil, err
 	}
-	if err := wrap.afterEnvFile.Close(); err != nil {
-		return nil, fmt.Errorf("failed to close after env file: %w", err)
-	}
 
 	absolutePathToHook, err := filepath.Abs(wrap.hookPath)
 	if err != nil {
 		return nil, fmt.Errorf("finding absolute path to %q: %w", wrap.hookPath, err)
 	}
 
-	tmplInput := scriptTemplateInput{
-		ShebangLine:       shebang,
-		BeforeEnvFileName: wrap.beforeEnvFile.Name(),
-		AfterEnvFileName:  wrap.afterEnvFile.Name(),
-		PathToHook:        absolutePathToHook,
-	}
-
-	// Create the hook runner code
-	buf := &strings.Builder{}
-	switch {
-	case isWindows && !isPOSIXHook && !isPwshHook:
-		batchScriptTmpl.Execute(buf, tmplInput)
-
-	case isWindows && isPwshHook:
-		powershellScriptTmpl.Execute(buf, tmplInput)
-
-	default:
-		posixShellScriptTmpl.Execute(buf, tmplInput)
-	}
-	script := buf.String()
-
-	// Write the hook script to the runner then close the file so we can run it
-	_, err = wrap.scriptFile.WriteString(script)
-	if err != nil {
+	if wrap.wrapperPath, err = WriteHookWrapper(
+		tmpl,
+		WrapperTemplateInput{
+			ShebangLine:       shebang,
+			BeforeEnvFileName: wrap.beforeEnvPath,
+			AfterEnvFileName:  wrap.afterEnvPath,
+			PathToHook:        absolutePathToHook,
+		},
+		scriptFileName,
+	); err != nil {
 		return nil, err
 	}
 
-	// Make script executable
-	err = utils.ChmodExecutable(wrap.scriptFile.Name())
-	if err != nil {
-		return nil, err
-	}
+	return wrap, nil
+}
 
-	// the defered close attempt will discard any errors,
-	// and attempting to execute a unclosed script will fail with ETXTBSY
-	if err := wrap.scriptFile.Close(); err != nil {
-		return nil, fmt.Errorf("failed to close script file: %w", err)
+func WriteHookWrapper(
+	tmpl *template.Template,
+	input WrapperTemplateInput,
+	hookWrapperName string,
+) (name string, err error) {
+	f, err := tmpfile.KeepExtensionWithMode(hookWrapperDir, hookWrapperName, 0o700)
+	if err != nil {
+		return "", err
 	}
-	return wrap, nil
+	defer func() {
```

Comment:
I think deferring the call to close that checks the error, and returning at `tmpl.Execute`, obscures the main flow, which is:

1. create the file
2. write to the file
3. close the file

Deferring `Close` as cleanup that happens in error flows makes sense. If `Close` did something like panic on a second call, then I can see this being useful.

Also "other: <nil>" would look weird...

So I think this should look more like:

```go
f, err := tmpfile.KeepExtensionWithMode(...)
if err != nil {
    return "", err
}
defer f.Close()
if err := tmpl.Execute(f, input); err != nil {
    return "", err
}
if err := f.Close(); err != nil {
    return "", err
}
return f.Name(), nil
```

Reply:
Ok, I've done this.